src/vcpkg/visualstudio.cpp

```diff
@@ -126,45 +126,39 @@ namespace vcpkg::VisualStudio
-            else
-                instances.emplace_back(
-                    common7_tools.parent_path().parent_path(), "16.0", VisualStudioInstance::ReleaseType::LEGACY);
-        }
-
-        const auto append_if_has_cl_vs140 = [&](fs::path&& path_root) {
+        const auto append_if_has_cl = [&](fs::path&& path_root, CStringView version) {
             const auto cl_exe = path_root / "VC" / "bin" / "cl.exe";
             const auto vcvarsall_bat = path_root / "VC" / "vcvarsall.bat";
 
             if (fs.exists(cl_exe) && fs.exists(vcvarsall_bat))
-                instances.emplace_back(std::move(path_root), "14.0", VisualStudioInstance::ReleaseType::LEGACY);
+                instances.emplace_back(
+                    std::move(path_root), version.c_str(), VisualStudioInstance::ReleaseType::LEGACY);
         };
 
-        // VS2015 instance from environment variable
-        auto maybe_vs140_comntools = System::get_environment_variable("vs140comntools");
-        if (const auto path_as_string = maybe_vs140_comntools.get())
-        {
-            // We want lexically_normal(), but it is not available
-            // Correct root path might be 2 or 3 levels up, depending on if the path has trailing backslash.
-            auto common7_tools = fs::u8path(*path_as_string);
-            if (common7_tools.filename().empty())
-                append_if_has_cl_vs140(common7_tools.parent_path().parent_path().parent_path());
-            else
-                append_if_has_cl_vs140(common7_tools.parent_path().parent_path());
-        }
+        const auto append_if_comntools_has_cl = [&](ZStringView env_var, CStringView version) {
+            auto maybe_comntools = System::get_environment_variable(env_var);
+            if (const auto path_as_string = maybe_comntools.get())
+            {
+                // We want lexically_normal(), but it is not available
+                // Correct root path might be 2 or 3 levels up, depending on if the path has trailing backslash.
+                auto common7_tools = fs::u8path(*path_as_string);
+                if (common7_tools.filename().empty())
+                    append_if_has_cl(common7_tools.parent_path().parent_path().parent_path(), version);
+                else
+                    append_if_has_cl(common7_tools.parent_path().parent_path(), version);
+            }
+        };
+
+        const auto maybe_append_legacy_vs = [&](ZStringView env_var, const fs::path& dir, CStringView version) {
+            // VS instance from environment variable
+            append_if_comntools_has_cl(env_var, version);
+            // VS instance from Program Files
+            append_if_has_cl(program_files_32_bit / dir, version);
+        };
 
-        // VS2015 instance from Program Files
-        append_if_has_cl_vs140(program_files_32_bit / "Microsoft Visual Studio 14.0");
+        append_if_comntools_has_cl("vs160comntools", "16.0");
```

Comment:
This is incorrect -- due to layout changes in VS 2017 (as part of the new installer), the compiler can no longer be located via a simple path navigation. This needs to be changed back to the previous approach (always add it after normalization).

A big difference between the different `VSXYZCOMNTOOLS` vars is that `vs160comntools` is actually not set system-wide as part of installation; it's only available as part of a developer prompt. This means its presence indicates "intent" on the developer's part to make this copy of VS available and it significantly reduces the burden of a false positive.

Reply:
That explains why the code looked duplicated to me before - I missed that detail. I hope you don't mind that I tried to keep things generic. It could prove useful if support is ever desired for even older toolsets.